server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java

```diff
@@ -207,6 +261,79 @@ public Iterator<Setting<?>> settings() {
+     * Also validates that if the low max headroom is set, then the high max headroom must be set as well.
+     * Also validates that if the high max headroom is set, then the flood stage max headroom must be set as well.
+     * Else, throws an exception.
+     */
+    static class MaxHeadroomValidator implements Setting.Validator<ByteSizeValue> {
+
+        @Override
+        public void validate(ByteSizeValue value) {
+
+        }
+
+        @Override
+        public void validate(final ByteSizeValue value, final Map<Setting<?>, Object> settings) {
+            final ByteSizeValue lowHeadroom = (ByteSizeValue) settings.get(CLUSTER_ROUTING_ALLOCATION_LOW_DISK_MAX_HEADROOM_SETTING);
+            final ByteSizeValue highHeadroom = (ByteSizeValue) settings.get(CLUSTER_ROUTING_ALLOCATION_HIGH_DISK_MAX_HEADROOM_SETTING);
+            final ByteSizeValue floodHeadroom = (ByteSizeValue) settings.get(
+                CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_MAX_HEADROOM_SETTING
+            );
+
+            if (lowHeadroom.getBytes() > 0 && highHeadroom.getBytes() < 0) {
+                throw new IllegalArgumentException(
+                    "high disk max headroom ["
+                        + highHeadroom.getStringRep()
+                        + "] is not set, while the low disk max headroom is set ["
+                        + lowHeadroom.getStringRep()
+                        + "]"
+                );
+            }
+            if (highHeadroom.getBytes() > 0 && floodHeadroom.getBytes() < 0) {
+                throw new IllegalArgumentException(
+                    "flood disk max headroom ["
+                        + floodHeadroom.getStringRep()
+                        + "] is not set, while the high disk max headroom is set ["
+                        + highHeadroom.getStringRep()
+                        + "]"
+                );
+            }
+
+            // For the comparisons, we need to mind that headroom values can default to -1.
+            if (highHeadroom.compareTo(lowHeadroom) > 0 && lowHeadroom.getBytes() > 0) {
+                throw new IllegalArgumentException(
+                    "high disk max headroom ["
+                        + highHeadroom.getStringRep()
+                        + "] more than low disk max headroom ["
+                        + lowHeadroom.getStringRep()
+                        + "]"
+                );
+            }
+            if (floodHeadroom.compareTo(highHeadroom) > 0 && highHeadroom.getBytes() > 0) {
+                throw new IllegalArgumentException(
+                    "flood disk max headroom ["
+                        + floodHeadroom.getStringRep()
+                        + "] more than high disk max headroom ["
+                        + highHeadroom.getStringRep()
+                        + "]"
+                );
+            }
+        }
+
+        @Override
+        public Iterator<Setting<?>> settings() {
+            final List<Setting<?>> settings = List.of(
```

Comment:
Well if I move it as is to a static final field I get:

```
Caused by: java.lang.NullPointerException
	at java.base/java.util.Objects.requireNonNull(Objects.java:208)
	at java.base/java.util.ImmutableCollections.listFromTrustedArray(ImmutableCollections.java:213)
	at java.base/java.util.List.of(List.java:847)
	at org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings$WatermarkValidator.<clinit>(DiskThresholdSettings.java:253)
```

Seems that the settings are not yet properly initialized. I can make it a static field and just assign it once in case it's null. I did that. I do not think we should have any concurrency correctness issues since they would just write the same list. Please tell me if you like this solution and resolve the conversation.

Reply:
I see. The static field you added looks like it is in `WatermarkValidator`, which makes this a cyclic initialization. If you add it as a static field directly on `DiskThresholdSettings` instead it should work fine.

> I do not think we should have any concurrency correctness issues since they would just write the same list.

I'd prefer not to do this. I think it works because `List.of` returns `ListN`, which has only final fields. But this is not documented and could in principle change for newer releases of java. If the object published has non-final fields, you risk users seeing a half-initialized object.

Reply:
Thank you @henningandersen , indeed I could do it as you said! Done and pushed. Feel free to unresolve this conversation if you feel it's not handled sufficiently.